### DISCUSSION TITLE
What is the best practice for integrating Okta, AWS SSO, and Gruntwork?

### DISCUSSION THREAD

**Discussion by pete0emerson** [gruntwork-io/knowledge-base#308] [Help]
We’ve gotten Okta and we are starting to integrate all of our SSO capable apps with it. The design for AWS came up and I noted that our choices there have significant impact on our ability to do automation with Grunwork’s Reference Library.
I would like your advice on how to do this. I would like to do as much as possible with Terraform to facilitate managed changes to the configuration going forward.

**Comment by antonl** [comment]
I've been able to integrate Okta and the RefArch by configuring Okta as a delegated IdP in AWS SSO and creating permission sets in the `security` account that grant permissions equivalent to the ref-arch security groups. This is a bit of a hack, but does work quite well provided you use aws-vault role chaining. You don't need any user accounts at all (except as a backup), nor do you need to change the ref-arch configurations.

For example, an inline ops-admin-access permission set could look like this: 
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "CoreAccounts",
            "Effect": "Allow",
            "Action": [
                "sts:TagSession",
                "sts:AssumeRole"
            ],
            "Resource": [
                "arn:aws:iam::{SECURITY_ID}:role/allow-ops-admin-access-from-other-accounts",
                "arn:aws:iam::{LOGS_ID}:role/allow-ops-admin-access-from-other-accounts",
                "arn:aws:iam::{SHARED_ID}:role/allow-ops-admin-access-from-other-accounts",
                "arn:aws:iam::{DEV_ID}:role/allow-ops-admin-access-from-other-accounts",
                "arn:aws:iam::{STAGE_ID}:role/allow-ops-admin-access-from-other-accounts",
                "arn:aws:iam::{PROD_ID}:role/allow-ops-admin-access-from-other-accounts"
            ]
        }
}
```

In order to configure aws-vault for this, you have to use role chaining. In `~/.aws/config` put snippets like this:
```
[profile allow-ops-admin-access]
sso_start_url=https://{SSO_DOMAIN}.awsapps.com/start
sso_region=us-west-2
sso_account_id={SECURITY_ID}
sso_role_name=allow-ops-admin-access

[profile logs-ops-admin]
source_profile=allow-ops-admin-access
role_arn=arn:aws:iam::{LOGS_ID}:role/allow-ops-admin-access-from-other-accounts
region=us-west-2
```

You can similar roles for the other accounts based on the permission sets, or parametrize your permission sets by account id or through SAML assertions.
